test/worlds.jl

```diff
@@ -199,3 +199,144 @@ end
+f(::Float64) = 2
+applyf([1])
+applyf([1.0])
+applyf(Any[1])
+wint   = worlds(instance(applyf, (Vector{Int},)))
+wfloat = worlds(instance(applyf, (Vector{Float64},)))
+wany2  = worlds(instance(applyf, (Vector{Any},)))
+src2 = code_typed(applyf, (Vector{Any},))[1]
+f(::String) = 3
+applyf(Any[1])
+@test worlds(instance(applyf, (Vector{Int},))) == wint
+@test worlds(instance(applyf, (Vector{Float64},))) == wfloat
+wany3 = worlds(instance(applyf, (Vector{Any},)))
+src3 = code_typed(applyf, (Vector{Any},))[1]
+@test (wany3 == wany2) == equal(src3, src2)   # don't invalidate unless you also change the code
+f(::AbstractVector) = 4         # next test would pass if this were ::Vector{Int}
+applyf(Any[1])
+wany4 = worlds(instance(applyf, (Vector{Any},)))
+src4 = code_typed(applyf, (Vector{Any},))[1]
+@test_broken (wany4 == wany3) == equal(src4, src3)
+f(::Dict) = 5
+applyf(Any[1])
+wany5 = worlds(instance(applyf, (Vector{Any},)))
+src5 = code_typed(applyf, (Vector{Any},))[1]
+@test (wany5 == wany4) == equal(src5, src4)
+f(::Set) = 6    # with current settings, this shouldn't invalidate
+applyf(Any[1])
+wany6 = worlds(instance(applyf, (Vector{Any},)))
+src6 = code_typed(applyf, (Vector{Any},))[1]
+@test (wany6 == wany5) == equal(src6, src5)
+
+## ambiguities do not trigger invalidation
+using Printf
+Printf.gen("%f")
+mi = instance(+, (AbstractChar, UInt8))
+w = worlds(mi)
+
+abstract type FixedPoint{T <: Integer} <: Real end
+struct Normed <: FixedPoint{UInt8}
+    i::UInt8
+    Normed(i::Integer, _) = new(i % UInt8)
+end
+(::Type{X})(x::Real) where X<:FixedPoint{T} where T = X(round(T, typemax(T)*x), 0)
+
+@test_broken worlds(mi) == w
+
+## specialization heuristics do not trigger false invalidation
+## Here, the new method is not actually reachable, but partial specialization makes the MethodInstance
+## specTypes make it look like it might be.
+s = Set([VERSION])
+mv = maximum(s)
+mi = instance(maximum, (typeof(s),))
+w = worlds(mi)
+Base.reduce_empty(::typeof(Base.add_sum), ::Type{F}) where {F<:FixedPoint}  = 0.0
+@test sum(Normed[]) === 0.0
+@test_broken worlds(mi) == w
+
+## Check that invalidations don't happen for mi.specTypes <: method.sig
+## In this case the MethodInstance being invalidated is for Union{}(::Array{_A,1} where _A)
+createdict(@nospecialize(d::Dict{K,V} where {K,V})) = typeof(d)()
+d = createdict(Dict("a"=>1))
+mi = instance(createdict, (Dict{K,V} where {K,V},))
+w = worlds(mi)
+struct SHermitianCompact{N,T,L}
+    data::NTuple{L,T}
+    SHermitianCompact{N,T,L}(data::NTuple{L,T}) where {N,T,L} = new{N,T,L}(data)
+end
+@inline (::Type{SSC})(a::AbstractVector) where {SSC <: SHermitianCompact} = SSC((a...,))
+x = SHermitianCompact{1,Float32,3}([1.0f0, 2.0f0, 3.0f0])
+@test_broken worlds(mi) == w
+
+## A case in which there is no intersection between signatures
+## A possibly relevant point is that the invalidated instance comes from
+##    convert(::Type{T}, x) where {T>:Nothing} = convert(nonnothingtype_checked(T), x)
```

Comment:
In this case `nonnothingtype_checked(T)` returns `Union{}`, so this calls `convert(Union{}, ::String)` which is ambiguous, so hopefully it's just another case of the ambiguity issue.